src/LEDs.c

```diff
@@ -0,0 +1,25 @@
+#include "stm32f4xx.h"
+#include "FreeRTOS.h"
+#include "FreeRTOSConfig.h"
+#include "task.h"
+
+static void blinkTask(void){
+	while(1){
+		GPIOD->ODR ^= GPIO_ODR_ODR_13;
+		vTaskDelay(200);

```

Comment:
Two things to say if you want the LED to blink at the speed you want:

1. vTaskDelay delays the task for a certain amount of ticks, not milliseconds. If you want to specify in milliseconds, you can do `vTaskDelay(200 / portTICK_PERIOD_MS)`. So, you saying it blinks slower than intended probably means that one tick is a few milliseconds.
2. I know in the rest of the code we always used vTaskDelay, and since we only execute one line of code in the loop, it works pretty well. However, vTaskDelay delays you from the moment you call it, so your actual task period is: 200 ticks + however long it takes to finish executing the GPIOD->ODR line of code. This time can increase randomly if you have bad timing with interrupts or other tasks preempting you in between, though that wouldn't be super likely seeing as you're only executing one line of code. If you wanted to ensure that the LED was more consistent in blinking in 200 tick intervals, then you can use vTaskDelayUntil, which takes in as an input the previous time at which it woke up. That way, if your one line of code takes, for example, 6 ticks to execute, then vTaskDelayUntil will only sleep 194 ticks, instead of 200. This would mean that your overall task execution period will be very consistently 200 ticks. Now, this isn't super crucial for something like an LED, but it might be worth thinking about for other tasks if ever you write a task for a more time-critical component, with more stuff going on in its execution loop.

Hope this helped!

Reply:
Thanks for you comment @lymacasm . We tried using the pdMS_TO_TICKS macro but it gives the same result. The RTOS scheduler does seem to be configured to run every millisecond, but it seems that is not actually happening. Maybe the micro isn't actually running at the clock speed it is configured at. It should be running at 168MHz. We are using the STM32F407 for this project. Do you know of a way to measure the clock speed on the micro?

Reply:
Sorry for the late reply! You can setup a GPIO pin for PWM with 50% duty cycle and the smallest period possible. You can then use an oscilloscope to measure the frequency, double it, and you have the frequency of your RTC/GPIO clock. This should hopefully be the same as your SysTick frequency (the ARM timer interrupt used by FreeRTOS).

If the frequency you get doesn't make sense, the STM boards has a bunch of prescaling options on their timers. If you really want to be sure that the frequency you read is correct, you can look into the timer routing and do the math for all the registers. You'll want to find the clock frequency for the clock signal going into the CPU, as that signal will be used to generate you'll beloved SysTick interrupts.

Let me know if you have any questions! I hope that helped!